timur/lib/client/jsx/components/browser/browser.jsx

```diff
@@ -67,21 +77,62 @@ export default function Browser({model_name, record_name, tab_name}) {
   );
   const {view, record, tab, revision, template, can_edit} = browserState;
   const [mode, setMode] = useState('loading');
-  const loading = !view || !template || !record || !tab_name;
+  const [error, setError] = useState(null);
   const {cancelEdits, approveEdits} = useEditActions(setMode, browserState);
-  const {selectOrShowTab, selectDefaultTab, selectTab, showTab} = useTabActions(
-    browserState,
-    setMode
-  );
+  const requestDocuments = useRequestDocuments();
+
+  const browseToTab = useCallback((tabName) => {
+    invoke(setLocation(window.location.href.replace(/#.*/, '') + `#${tabName}`));
+  }, [invoke, setLocation]);
 
   // Set at 'skin' on the browser styling.
   let skin = 'browser';
   if (mode === 'browse') skin = 'browser ' + model_name;
   const editMode = useCallback(() => setMode('edit'), [setMode]);
 
-  // On mount
+  // const loading = !view || !template || !record || !tab_name;
+  const [loading, loadDocuments] = useAsyncWork(function* loadDocuments(model_name, view) {
+    setMode('loading');
+
+    if (!template) yield invoke(requestModel(model_name));
+    if (!view) {
+      // we are told the model and record name, get the view
+      view = yield invoke(requestView(model_name));
+    }
+
+    if (!tab_name) {
+      browseToTab(getDefaultTab(view));
+      return;
```

Comment:
This is also a redirect after loading has commenced - it is a bit strange that this happens here while the other redirect happens in useEffect()

Reply:
The first browseTo that happens in useEffect only requires loading the project.  This browseTo requires the view to have loaded to determine the default tab, which cannot happen in the useEffect without also increasing complexity of that useEffect logic.

That said I can flatten this into the loader generator and simplify how it looks.

Reply:
Ok, check again, looks a lot more clean.